### ISSUE TITLE
(Re-)implement Gaussian likelihood

### ISSUE THREAD

**Issue by matsen** [matsengrp/torchdms#14]
This has several components.

## 1. Generalize data loading and processing

Measurements sometimes come with an estimate of their variance. For example if we are interested in a functional score, we would have an equal-sized vector describing the variance of each functional score measurement.

I should also mention that we are going to have stability data soon as well, along with its measurements of variance.

How shall we store this sometimes-present information?  We [are already abusing BinaryMap](https://github.com/matsengrp/torchdms/blob/master/torchdms/data.py#L82) by stuffing extra things in it.  We could either [add extra columns](https://matsengrp.slack.com/archives/CS62Y7CRL/p1586881560006100?thread_ts=1586866612.003500&cid=CS62Y7CRL) to BinaryMap, or we could come up with our own data structure.

@jgallowa07 , if we were going to use an xarray Dataset, would we come up with an object or just use a Dataset on its own? I assume we could store the one-hot matrix, various floating point data, and a string vector (the mutations themselves)?
 

## 2. Generalize training setup

Right now, the data expected for the training algorithm [is hard-coded in to the `train` method](https://github.com/matsengrp/torchdms/blob/master/torchdms/analysis.py#L56). I think that we should be able to replace the `criterion` argument with a function that takes a `batch` and a `device` and produces a loss.  Then we can call `train` with any objective on whatever data is supplied in the batch.

We also need to generalize the current `BinarymapDataset` to several versions that supply [different versions of the data as needed](https://github.com/matsengrp/torchdms/blob/master/torchdms/data.py#L34).  I guess we can have an ABC and then a bunch of subclasses defining different `__getitem__` methods?

## 3. Actually implement a Gaussian likelihood

@ZorianThornton5 implemented this like so:

```python

def gaus_loglik(inputs, mu, var):
    """ Find the current log-likelihood of Gaussian distribution
    
    Args:
        - inputs (tensor): Input labels (nx1 tensor)
        - mu (tensor): Target to estimate (px1 tensor)
        - var (tensor): Variance(s) of distribution/variants (1x1 tensor)
        
    Returns:
        - loglik (tensor): Float value of current log likelihood
    """
    pi = torch.Tensor([math.pi]) # Define pi
    loglik = torch.sum( (-torch.log(2*pi)/2) - torch.log(torch.sqrt(var)) - (inputs - mu)**2/2*var)
    return -loglik
```

Note that in our case, `mu` is the observation we want to fit, and `var` is the variance (also supplied by the data). 

This may be the way to go, though if it works I would prefer to use PyTorch's [Normal](https://pytorch.org/docs/stable/distributions.html#torch.distributions.normal.Normal) distribution that is equipped with a [log_prob](https://pytorch.org/docs/stable/distributions.html#torch.distributions.distribution.Distribution.log_prob) method that I can only imagine is back-proppable and vectorized.

## 4. Work on language

We are going to have "variants" and "variance" 😱 . What to do?????????????????????
